### PR TITLE
Re-open the ldif file to prevent error message

### PR DESCRIPTION
**Pull request by Tiboris**
There was an issue with ipa-server-upgrade and it was
showing an error while upgrading:
DN... does not exists or haven't been updated.

Resolves: https://pagure.io/freeipa/issue/7644

### CODE REVIEW

File: ipaserver/install/upgradeinstance.py [hunk 1]
```diff
@@ -241,6 +241,9 @@ def __disable_schema_compat(self):
             with open(self.filename, "r") as in_file:
                 parser = GetEntryFromLDIF(in_file, entries_dn=[COMPAT_DN])
                 parser.parse()
+
+                # move to file begining because parser is reading same file
```
[the remark this answers is not in the document]
Comment (Tiboris) [reply ipaserver/install/upgradeinstance.py]: Of course, thanks for explanation.

**Comment by flo-renaud** [comment]
Hi @Tiboris 
thank you for the patch, works for me. I tested with fresh install + ipa-server-upgrade and the error message was not present any more.
As there is already a test case in `ipatests/test_integration/test_upgrade.py`, you can easily add enhance the existing test by checking that the string `DN: cn=Schema Compatibility,cn=plugins,cn=config does not exists or haven't been updated` is not in cmd.stdout_text (or result.stderr_text).

**Comment by Tiboris** [comment]
Thanks @flo-renaud will do.

**Comment by tiran** [comment]
IMHO the method would benefit from some refactorization. For example it creates ``ldif_outfile`` even if ``COMPAT_DN`` is missing from the infile.

```
        with open(self.filename, "r") as in_file:
            parser = GetEntryFromLDIF(in_file, entries_dn=[COMPAT_DN])
            parser.parse()

        try:
            compat_entry = parser.get_results()[COMPAT_DN]
        except KeyError:
            return

        with open(ldif_outfile, "w") as out_file:
            with open(self.filename, "r") as in_file:
                parser = installutils.ModifyLDIF(in_file, out_file)
                ...
```

**Comment by Tiboris** [comment]
Thanks for the tip @tiran!

File: ipaserver/install/upgradeinstance.py [hunk 2]
```diff
@@ -237,14 +237,18 @@ def __disable_listeners(self):
...
         with open(ldif_outfile, "w") as out_file:
             with open(self.filename, "r") as in_file:
-                parser = GetEntryFromLDIF(in_file, entries_dn=[COMPAT_DN])
-                parser.parse()
-                try:
-                    compat_entry = parser.get_results()[COMPAT_DN]
-                except KeyError:
-                    return
                 parser = installutils.ModifyLDIF(in_file, out_file)
                 if not compat_entry.get('nsslapd-pluginEnabled'):
```
Comment (flo-renaud) [inline ipaserver/install/upgradeinstance.py]: You can probably move this check before open(ldif_outfile,...). This way, if the input file does not contain the entry nssldap-pluginEnabled, the outfile doesn't get created and we exit quickly.

Comment (Tiboris) [reply ipaserver/install/upgradeinstance.py]: I will try!

Comment (Tiboris) [reply ipaserver/install/upgradeinstance.py]: Thanks @flo-renaud, 
ipa-server-upgrade is working for me with this change.

**Comment by rcritten** [comment]
You can fix this as a one-liner with:

in_file.seek(0) before the second invocation of parser = installutils.ModifyLDIF(in_file, out_file)

**Comment by Tiboris** [comment]
Hello @rcritten,

thanks for input. Original pull request used your approach. On the other side I agree with @tiran, as he showed me that this code could look better and this PR created a opportunity to do so. Yes it is maybe opening same file twice and maybe this method could have been more lightweight calling `open()` only once. Idea of improving code that it does not create any artifacts got me but I will not insist on the shape of PR as it is.

**Comment by rcritten** [comment]
Ok, it LGTM.

**Comment by Tiboris** [comment]
Maybe some changes in `ModifyLDIF()` could create better image and resolve this issue with re-reading file.

**Comment by rcritten** [comment]
IMHO it is fine as-is. This is only executed during installs and on upgrades so re-reading the file has limited impact.

**Comment by flo-renaud** [comment]
Hi @Tiboris 
thanks for the patch and the test, works for me.
Tested by fresh install + ipa-server-upgrade, the msg does not appear any more.

**Comment by flo-renaud** [comment]
master:

* 6fa1e6f18ef1798fa8cd5030807c81699fdfbdc6 Re-open the ldif file to prevent error message
* 89799a14ce674f73bdfb8310256e7c8c67866a34 Add assert to check output of upgrade

**Pull request closed without merging**